### DISCUSSION TITLE
Can't open connection to MS SQL server in container

### DISCUSSION THREAD

**Discussion by Polivando** [testcontainers/testcontainers-dotnet#539] [Q&A]
Hi
So I'm trying to set up unit tests that talk to my database running in docker container. I used this library, and for now it's mostly code copied from [this file](https://github.com/testcontainers/testcontainers-dotnet/blob/0f4d0ff3ab79567bda94b8d3d41e0a9987665497/tests/Testcontainers.Tests/Unit/Containers/Unix/Modules/Databases/MsSqlTestcontainerTest.cs#L13) and corresponding fixture.

I am able to start the container up, I see logs indicating that DB is running, I can execute scripts using `sqlcmd `command, but the ConnectionEstablished test keeps failing and I can't understand why.

Can't connect to the DB via ms sql management studio as well.

I'm positive I'm using correct port# when connecting through management studio.
I'm on Windows with WSL, and I have docker installed as part of Rancher Desktop.

What can be the reason of connection failing to establish and what should I check/change to fix this issue?

**Comment by Polivando** [comment]
[Test.DockerDb.Tests.zip](https://github.com/testcontainers/testcontainers-dotnet/files/9240926/Test.DockerDb.Tests.zip)
Here's the project I'm using

**Reply by HofmeisterAn** [reply]
FYI: You can reply on answers instead of creating a new one. Your example looks good and runs fine on my machine WSL + Docker Desktop.

**Reply by Polivando** [reply]
yeah, well, the ConnectionEstablished fails for me.
could this be because I didn't install docker desktop separately, but as part of Rancher Desktop?
My enterprise does not meet the requirements to use docker desktop for free for commercial use, and I don't have clearance to use the paid version, so I installed Rancher, which is free

**Reply by HofmeisterAn** [reply]
I do not know much about this setup, but did you try to connect to a container running from the CLI? How does the connection string look like?

**Reply by Polivando** [reply]
yes, the thing is that cli connection (via sqlcmd) is the only way to connect that actually works
connection via testcontainers fails, and I'm also not able to connect via SQL management studio
connections string for Management studio looks like: `localhost:{port that I get from containers list}`, username = SA, password is what's hardcoded in fixture. Also tried with 127.0.0.1 instead of localhost

CLI command is `sqlcmd -S 127.0.0.1,49155 -U sa -P 'yourStrong(!)Password'`

**Reply by HofmeisterAn** [reply]
Did you try the gateway IP address?

**Comment by ArwynFr** [comment]
Hello, you might be interested in this SO article, the problem is linked to ipv6:

> [Docker in WSL - Can't connect to SQL server on 'localhost' but with '127.0.0.1' SQL Server 2017-latest](https://stackoverflow.com/questions/75446957/docker-in-wsl-cant-connect-to-sql-server-on-localhost-but-with-127-0-0-1)

You can either use 127.0.0.1 instead of localhost, or disable ipv6

**Reply by ArwynFr** [reply]
Interestingly, it looks like [::1] works, so it's not IPv6 based.

I found an ms technet discussion archive stating that localhost uses Shared Memory instead of TCP/IP:
https://learn.microsoft.com/en-us/archive/msdn-technet-forums/7e32b20c-a5dc-456a-91ae-23bf121784a4
So basically there is zero chance that using localhost would work if you're on a container database with a windows client.

I also found that testcontainers will use the `DOCKER_HOST` hostname to generate the connection string.
I was able to make it work by replacing `tcp://localhost:2375` by `tcp://[::1]:2375` to access my wsl docker engine.
This in turn made testcontainers to use `Server=[::1],<port>` in the default connection string.
The application then used a TCP/IP connection which succeeded in accessing the database !
